daemon/containerd/image_list.go

```diff
@@ -79,51 +82,61 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 		layers = make(map[digest.Digest]int)
 	}
 
+	platform := platforms.AllPlatformsWithPreference(cplatforms.Default())
+	// TODO allow platform override?
+
 	contentStore := i.client.ContentStore()
 	for _, img := range imgs {
 		if !filter(img) {
 			continue
 		}
 
+		// TODO this logic matches GetImage and should probably be combined somehow (although we keep the ImageManifest object)
+		type tempImage struct {
+			img *ImageManifest
+			oci *ocispec.Image
+		}
+		var presentImages []tempImage
 		err := i.walkImageManifests(ctx, img, func(img *ImageManifest) error {
-			if isPseudo, err := img.IsPseudoImage(ctx); isPseudo || err != nil {
-				return err
-			}
-
-			available, err := img.CheckContentAvailable(ctx)
-			if err != nil {
-				log.G(ctx).WithFields(logrus.Fields{
-					logrus.ErrorKey: err,
-					"manifest":      img.Target(),
-					"image":         img.Name(),
-				}).Warn("checking availability of platform specific manifest failed")
-				return nil
-			}
-
-			if !available {
-				return nil
-			}
-
-			image, chainIDs, err := i.singlePlatformImage(ctx, contentStore, img)
+			conf, err := img.Config(ctx)
 			if err != nil {
 				return err
 			}
-
-			summaries = append(summaries, image)
-
-			if opts.SharedSize {
-				root = append(root, &chainIDs)
-				for _, id := range chainIDs {
-					layers[id] = layers[id] + 1
-				}
+			var ociimage ocispec.Image
+			if err := readConfig(ctx, contentStore, conf, &ociimage); err != nil {
+				return err
 			}
-
+			presentImages = append(presentImages, tempImage{img, &ociimage})
 			return nil
 		})
 		if err != nil {
 			return nil, err
 		}
+		if len(presentImages) == 0 {
+			// TODO we should probably show *something* for images we've pulled but are 100% shallow or an empty manifest list/index ("tianon/scratch:index" is an empty example image index and "tianon/scratch:list" is an empty example manifest list)
```

Comment:
Unrelated to listing images, should we even pull these images? Or rather shouldn't we fail and delete images like these? Since we know we can't run them.